op-challenger/game/keccak/merkle/tree.go

```diff
@@ -0,0 +1,193 @@
+package merkle
+
+import (
+	"errors"
+	"math/big"
+
+	"github.com/ethereum-optimism/optimism/op-challenger/game/keccak/types"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+// BinaryMerkleTreeDepth is the depth of the merkle tree.
+const BinaryMerkleTreeDepth = 16
+
+var (
+	// MaxLeafCount is the maximum number of leaves in the merkle tree.
+	MaxLeafCount = 1<<BinaryMerkleTreeDepth - 1 // 2^16 - 1
+	// IndexOutOfBoundsError is returned when an index is out of bounds.
+	IndexOutOfBoundsError = errors.New("index out of bounds")
+	// zeroHashes is a list of empty hashes in the binary merkle tree, indexed by height.
+	zeroHashes [BinaryMerkleTreeDepth]common.Hash
+)
+
+func init() {
+	// Initialize the zero hashes. These hashes are pre-computed for the starting state of the tree, where all leaves
+	// are equal to `[32]byte{}`.
+	for height := 0; height < BinaryMerkleTreeDepth-1; height++ {
+		zeroHashes[height+1] = crypto.Keccak256Hash(zeroHashes[height][:], zeroHashes[height][:])
+	}
+}
+
+// MerkleProof is a keccak merkle proof, which is a list of hashes that prove the inclusion of a leaf in the merkle tree.
+type MerkleProof [][]byte
+
+// toSized converts a [MerkleProof] to a list of commitment hashes.
+func ToSized(m MerkleProof) []common.Hash {
+	var sized []common.Hash
+	for _, proof := range m {
+		// SAFETY: if the proof is less than 32 bytes, it will be padded with 0s
+		if len(proof) < 32 {
+			proof = append(proof, make([]byte, 32-len(proof))...)
+		}
+		// SAFETY: the proof is 32 or more bytes here, so it will be truncated to 32 bytes
+		sized = append(sized, common.Hash(proof[:32]))
+	}
+	return sized
+}
+
+// merkleNode is a single node in the binary merkle tree.
+type merkleNode struct {
+	Label  common.Hash
+	GIndex *big.Int
```

Comment:
We don't appear to use `GIndex` for anything so probably worth removing it. Otherwise we need to add a bunch of tests to confirm it is calculated correctly.

Reply:
Good callout, agreed. Was legacy code I used but simplified out!